File: 997.Find-the-Town-Judge.go
// https://leetcode.com/problems/find-the-town-judge/
//
// algorithms
// Easy (46.92%)
// Total Accepted:    3,310
// Total Submissions: 7,055
// beats 100.0% of golang submissions

package leetcode

func findJudge(N int, trust [][]int) int {
	if N == 1 {
		return 1
	}
	arr := make([]int, N+1)
	judgeIdx := -1

	for _, t := range trust {
		arr[t[0]] = -1
		if arr[t[1]] == -1 {
			continue
		}
		arr[t[1]]++
		if arr[t[1]] == N-1 {
			judgeIdx = t[1]
		}
	}

	if judgeIdx > -1 && arr[judgeIdx] != -1 {
		return judgeIdx
	}

	return -1
}

File: 997.Find-the-Town-Judge.py
# https://leetcode.com/problems/find-the-town-judge/
# Easy (49.39%)
# Total Accepted:    32,301
# Total Submissions: 65,395
# beats 100.0% of python submissions


class Solution(object):
    def findJudge(self, N, trust):
        """
        :type N: int
        :type trust: List[List[int]]
        :rtype: int
        """
        trust_num = [0] * (N + 1)

        for a, b in trust:
            trust_num[a] -= 1
            trust_num[b] += 1

        for i in xrange(1, N + 1):

            if trust_num[i] == N - 1:
                return i

        return -1
